operator/duties/voluntary_exit.go

```diff
@@ -0,0 +1,99 @@
+package duties
+
+import (
+	"context"
+	"sync"
+
+	"github.com/attestantio/go-eth2-client/spec/phase0"
+	spectypes "github.com/bloxapp/ssv-spec/types"
+	"go.uber.org/zap"
+
+	"github.com/bloxapp/ssv/logging/fields"
+)
+
+const voluntaryExitSlotsToPostpone = phase0.Slot(4)
+
+type ExitDescriptor struct {
+	PubKey         phase0.BLSPubKey
+	ValidatorIndex phase0.ValidatorIndex
+	BlockNumber    uint64
+}
+
+type VoluntaryExitHandler struct {
+	baseHandler
+	validatorExitCh <-chan ExitDescriptor
+	dutyQueue       []*spectypes.Duty
+	dutyQueueMu     sync.Mutex
+}
+
+func NewVoluntaryExitHandler(validatorExitCh <-chan ExitDescriptor) *VoluntaryExitHandler {
+	return &VoluntaryExitHandler{
+		validatorExitCh: validatorExitCh,
+		dutyQueue:       make([]*spectypes.Duty, 0),
+	}
+}
+
+func (h *VoluntaryExitHandler) Name() string {
+	return spectypes.BNRoleVoluntaryExit.String()
+}
+
+func (h *VoluntaryExitHandler) HandleDuties(ctx context.Context) {
+	h.logger.Info("starting duty handler")
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-h.ticker.Next():
+			currentSlot := h.ticker.Slot()
+
+			h.logger.Debug("🛠 ticker event", fields.Slot(currentSlot))
+
+			var dutiesForExecution []*spectypes.Duty
+
+			h.dutyQueueMu.Lock()
+			for _, duty := range h.dutyQueue {
+				if duty.Slot > currentSlot {
+					break
```

Comment:
I think this should be `continue`, otherwise the subsequent duties that should be executed now would be ignored. I realize this may not happen because `validatorExitCh` should receive messages with increasing block numbers, however it's always safer to make less assumptions about external behavior, especially with channels (in some cases channels don't preserve order of messages).

What do you think about something like this?
```go
var forExecution []*spectypes.Duty
var remaining []*spectypes.Duty
for _, duty := range h.dutyQueue {
	if duty.Slot <= h.currentSlot {
		forExecution = append(forExecution, duty)
	} else {
		remaining = append(remaining, duty)
	}
}
h.dutyQueue = remaining
// ...
```

Reply:
> I'm not sure what could be the case when order of messages is not preserved, but okay, let's make it less error-prone.
For example, when multiple goroutines send to a channel, then the order of arrival is not guaranteed to match the order of who started sending first.

Reply:
Well, task executor is what triggers sending to `validatorExitCh`, so slots should be in order. Also, with the new approach we are going to copy duties every slot when they exist.

However, since this is a rare duty, the queue is expected to be small, and any bug here would be hard to find, I agree to change this.